File: Retaguarda/Torneio.Domain/Entities/CustoTorneio.cs
using Torneio.Domain.Enums;

namespace Torneio.Domain.Entities;

public class CustoTorneio
{
    public Guid Id { get; private set; }
    public Guid TorneioId { get; private set; }
    public CategoriaCustoTorneio Categoria { get; private set; }
    public string Descricao { get; private set; } = null!;
    public decimal Quantidade { get; private set; }
    public decimal ValorUnitario { get; private set; }
    public decimal ValorTotal { get; private set; }
    public DateTime? Vencimento { get; private set; }
    public string? Responsavel { get; private set; }
    public string? Observacao { get; private set; }

    private CustoTorneio() { }

    public static CustoTorneio Criar(
        Guid torneioId,
        CategoriaCustoTorneio categoria,
        string descricao,
        decimal quantidade,
        decimal valorUnitario,
        DateTime? vencimento,
        string? responsavel,
        string? observacao)
    {
        return new CustoTorneio
        {
            Id = Guid.NewGuid(),
            TorneioId = torneioId,
            Categoria = categoria,
            Descricao = descricao,
            Quantidade = quantidade,
            ValorUnitario = valorUnitario,
            ValorTotal = quantidade * valorUnitario,
            Vencimento = vencimento?.Date,
            Responsavel = string.IsNullOrWhiteSpace(responsavel) ? null : responsavel.Trim(),
            Observacao = string.IsNullOrWhiteSpace(observacao) ? null : observacao.Trim()
        };
    }

    public void Atualizar(
        CategoriaCustoTorneio categoria,
        string descricao,
        decimal quantidade,
        decimal valorUnitario,
        DateTime? vencimento,
        string? responsavel,
        string? observacao)
    {
        Categoria = categoria;
        Descricao = descricao;
        Quantidade = quantidade;
        ValorUnitario = valorUnitario;
        ValorTotal = quantidade * valorUnitario;
        Vencimento = vencimento?.Date;
        Responsavel = string.IsNullOrWhiteSpace(responsavel) ? null : responsavel.Trim();
        Observacao = string.IsNullOrWhiteSpace(observacao) ? null : observacao.Trim();
    }
}

File: App/torneio_app/lib/core/models/custo_torneio.dart
import 'package:intl/intl.dart';

class CustoTorneio {
  final String id;
  final String torneioId;
  final String categoria;
  final String descricao;
  final double quantidade;
  final double valorUnitario;
  final double valorTotal;
  final String categoriaLabel;
  final DateTime? vencimento;
  final String? responsavel;
  final String? observacao;
  final bool derivadoDaEmbarcacao;
  final String? equipeId;

  const CustoTorneio({
    required this.id,
    required this.torneioId,
    required this.categoria,
    required this.descricao,
    required this.quantidade,
    required this.valorUnitario,
    required this.valorTotal,
    required this.categoriaLabel,
    this.vencimento,
    this.responsavel,
    this.observacao,
    required this.derivadoDaEmbarcacao,
    this.equipeId,
  });

  factory CustoTorneio.fromJson(Map<String, dynamic> json) => CustoTorneio(
        id: json['id'] as String? ?? '',
        torneioId: json['torneioId'] as String,
        categoria: json['categoria'] as String? ?? 'Outros',
        descricao: json['descricao'] as String? ?? '',
        quantidade: (json['quantidade'] as num?)?.toDouble() ?? 0,
        valorUnitario: (json['valorUnitario'] as num?)?.toDouble() ?? 0,
        valorTotal: (json['valorTotal'] as num?)?.toDouble() ?? 0,
        categoriaLabel: json['categoriaLabel'] as String? ?? _labelCategoria(json['categoria'] as String?),
        vencimento: json['vencimento'] == null ? null : DateTime.tryParse(json['vencimento'].toString()),
        responsavel: json['responsavel'] as String?,
        observacao: json['observacao'] as String?,
        derivadoDaEmbarcacao: json['derivadoDaEmbarcacao'] as bool? ?? false,
        equipeId: json['equipeId'] as String?,
      );

  String get valorTotalFormatado => NumberFormat.currency(locale: 'pt_BR', symbol: 'R\$').format(valorTotal);

  static String _labelCategoria(String? categoria) => switch (categoria) {
        'Embarcacao' => 'Embarcação',
        'Camisas' => 'Camisas',
        'Alimentacao' => 'Alimentação',
        'Combustivel' => 'Combustível',
        'Premiacoes' => 'Premiações',
        _ => 'Outros',
      };
}
